### PR TITLE
chore: 16391 Optimize memory usage of the TssDirectory

### LINKED ISSUE #16391: `tss-library` Optimize memory usage by removing in directory `TssShareId`s classes

### PR DESCRIPTION
**Pull request by mxtartaglia-sl**
**Description**:
This Pr 
We aim to optimize the TssDirectory memory usage by removing the necessity of storing TssShareId instances.
Also, we simplify data structures that need to map an entity to a shareId using arrays and positional indexes.

**Related issue(s)**:

Fixes #16391

** Order of PR ** 
This pr depends on previous work, please also approve:
1: https://github.com/hashgraph/hedera-cryptography/pull/104
Prs depending on this one:
3: https://github.com/hashgraph/hedera-cryptography/pull/106
4: https://github.com/hashgraph/hedera-cryptography/pull/102

**Notes for reviewer**:
⚠️ Plaese review filtering commits: 16391

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: cryptography/hedera-cryptography-tss/src/main/java/com/hedera/cryptography/tss/api/TssParticipantDirectory.java [hunk 1]
```diff
@@ -56,64 +54,69 @@
...
-    private final Map<TssShareId, Long> shareAllocationMap;
+    private final List<Integer> ownedShareIds;
     /**
-     * Stores the {@link BlsPublicKey} of each {@code participantId} in the protocol.
+     * Stores the {@code participant} that is the owner of each shareId in the protocol.
+     * Index 0 represents shareId 1 and so on.
+     * Shares are assigned sequentially.
      */
-    private final Map<Long, BlsPublicKey> tssEncryptionPublicKeyMap;
+    private final Integer[] shareAllocationTable;
```
Comment (lpetrovic05) [inline cryptography/hedera-cryptography-tss/src/main/java/com/hedera/cryptography/tss/api/TssParticipantDirectory.java]: why not int[]?

File: cryptography/hedera-cryptography-tss/src/main/java/com/hedera/cryptography/tss/api/TssParticipantDirectory.java [hunk 2]
```diff
@@ -56,64 +54,69 @@
...
-                currentParticipantOwnedShareIds, "currentParticipantOwnedShareIds must not be null"));
-        this.shareAllocationMap =
-                Map.copyOf(Objects.requireNonNull(shareAllocationMap, "shareAllocationMap must not be null"));
-        this.tssEncryptionPublicKeyMap = Map.copyOf(
-                Objects.requireNonNull(tssEncryptionPublicKeyMap, "tssEncryptionPublicKeyMap must not be null"));
-        this.tssEncryptionPrivateKey =
-                requireNonNull(tssEncryptionPrivateKeyStore, "tssEncryptionPrivateKeyStore must not be null");
+        this.participantId = participantId;
+        this.shareIds = List.copyOf(shareIds);
+        this.ownedShareIds = List.copyOf(ownedShareIds);
```
Comment (lpetrovic05) [inline cryptography/hedera-cryptography-tss/src/main/java/com/hedera/cryptography/tss/api/TssParticipantDirectory.java:109]: why copy when this is a private constructor?

File: cryptography/hedera-cryptography-tss/src/main/java/com/hedera/cryptography/tss/api/TssParticipantDirectory.java [hunk 3]
```diff
@@ -56,64 +54,69 @@
...
-            @NonNull final List<TssShareId> currentParticipantOwnedShareIds,
-            @NonNull final Map<TssShareId, Long> shareAllocationMap,
-            @NonNull final Map<Long, BlsPublicKey> tssEncryptionPublicKeyMap,
-            @NonNull final PrivateKeyStore tssEncryptionPrivateKeyStore,
+            final int participantId,
+            @NonNull final List<Integer> shareIds,
+            @NonNull final List<Integer> ownedShareIds,
+            @NonNull final Integer[] shareAllocationTable,
+            @NonNull final BlsPublicKey[] tssEncryptionPublicKeyTable,
+            @NonNull final BlsPrivateKey tssDecryptionPrivateKey,
```
Comment (lpetrovic05) [inline cryptography/hedera-cryptography-tss/src/main/java/com/hedera/cryptography/tss/api/TssParticipantDirectory.java]: I feel like these data structures are a bit confusing, but I'm not sure if this is something that is worth changing

File: cryptography/hedera-cryptography-tss/src/main/java/com/hedera/cryptography/tss/api/TssParticipantDirectory.java (hunk 2 above)
Comment (mxtartaglia-sl) [reply cryptography/hedera-cryptography-tss/src/main/java/com/hedera/cryptography/tss/api/TssParticipantDirectory.java:109]: This ensures that no matter what list we receive in the constructor, it is always unmodifiable, and we can return the instance to the outside world without the risk of its content being changed.

File: cryptography/hedera-cryptography-tss/src/main/java/com/hedera/cryptography/tss/api/TssParticipantDirectory.java (hunk 3 above)
Comment (mxtartaglia-sl) [reply cryptography/hedera-cryptography-tss/src/main/java/com/hedera/cryptography/tss/api/TssParticipantDirectory.java]: They are, but I either solve the storage of this info with memory or efficiency.
At least, after this PR, handling these data structures is all the directory's responsibility, and no other piece needs to know how to navigate them.

File: cryptography/hedera-cryptography-tss/src/main/java/com/hedera/cryptography/tss/api/TssParticipantDirectory.java (hunk 1 above)
Comment (mxtartaglia-sl) [reply cryptography/hedera-cryptography-tss/src/main/java/com/hedera/cryptography/tss/api/TssParticipantDirectory.java]: It can definitely be.
Done

**Review by lpetrovic05 — APPROVED** [review]

**Review by andrewb1269 — APPROVED** [review]
Reviewed file:
`hedera-dependency-versions/build.gradle.kts`

Approved.

**Pull request merged**